c/authService.c

```diff
@@ -173,16 +188,209 @@ static int serveAuthCheck(HttpService *service, HttpResponse *res) {
     respondWithError(res, HTTP_STATUS_BAD_REQUEST, "Unexpected access level");
     return 0;
   }
-  /* printf("query: user %s, class %s, entity %s, access %d\n", userName, class,
-      entity, access); */
+  // printf("\n\nquery: user %s, class %s, entity %s, access %d accessStr %s\n", userName, class,
+      // entity, access, accessStr);
   privilegedServerName = getConfiguredProperty(service->server,
       HTTP_SERVER_PRIVILEGED_SERVER_PROPERTY);
   rc = zisCheckEntity(privilegedServerName, userName, class, entity, access,
       &reqStatus);
+      // printf("\n\nprivileged server name: %s", privilegedServerName);
   respond(res, rc, &reqStatus);
   return 0;
 }
 
+const char* getProfileNameFromRequest(char *url, char *method, int instanceID) {
+  char type[8]; // core || config || service
+  char productCode[1024];
+  char rootServiceName[1024];
+  char subUrl[15][1024];
+  char *profileName = (char*) safeMalloc(1024, "profileName");
+  char scope[1024];
+  char _p[1024], pluginID[1024], _s[1024], serviceName[1024], _v[1024];
+  char regexStr[] = "^/[A-Za-z0-9]*/plugins/";
+  
+  regex_t regex;
+  int value;
+  value = regcomp(&regex, regexStr, REG_EXTENDED);
+
+  if (value != 0) {
+    zowelog(NULL, LOG_COMP_ID_SECURITY, ZOWE_LOG_DEBUG2,
+           "RegEx compiled successfully.");
+  } else {
+    zowelog(NULL, LOG_COMP_ID_SECURITY, ZOWE_LOG_DEBUG2,
+           "RegEx compilation error %s.", regexStr);
+  }
+  value = regexec(&regex, url, 0, NULL, 0);
+  char urlCpy[1024];
+  snprintf(urlCpy, 1024, url);
+  int index = 0;
+  while (urlCpy[index]) { // Capitalize query
+      urlCpy[index] = toupper(urlCpy[index]);
+      index++;
+  }
+  if (instanceID < 0) { // Set instanceID
+    instanceID = 0;
+  }
+  if (value == REG_NOMATCH) {
+    zowelog(NULL, LOG_COMP_ID_SECURITY, ZOWE_LOG_DEBUG2,
+           "RegEx didn't match.");
+    char * token = strtok(urlCpy, "/");
+    int subUrlIndex = -1;
+    while( token != NULL ) {
+      if (strcmp(rootServiceName, NULL) == 0)
+      {
+        snprintf(rootServiceName, 1024, token);
+      } else {
+        snprintf(subUrl[subUrlIndex], 1024, token);
```

Comment:
What if you have more than 15 tokens? Buffer overflow?

Reply:
Made index check